File: rewrite-maven/src/test/java/org/openrewrite/maven/AddManagedDependencyTest.java
/*
 * Copyright 2022 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.maven;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.openrewrite.DocumentExample;
import org.openrewrite.Validated;
import org.openrewrite.test.RewriteTest;

import static org.assertj.core.api.Assertions.assertThat;
import static org.openrewrite.java.Assertions.mavenProject;
import static org.openrewrite.maven.Assertions.pomXml;

class AddManagedDependencyTest implements RewriteTest {

    @DocumentExample
    @Test
    void onlyAddedWhenUsing() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "2.17.2", "import",
            "pom", null,null, null, "org.apache.logging.log4j:*", false)),
          pomXml(
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencies>
                <dependency>
                  <groupId>org.apache.logging.log4j</groupId>
                  <artifactId>log4j-core</artifactId>
                  <version>2.17.2</version>
                </dependency>
              </dependencies>
            </project>
            """,
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>my-app</artifactId>
              <version>1</version>
              <dependencyManagement>
                <dependencies>
                  <dependency>
                    <groupId>org.apache.logging.log4j</groupId>
                    <artifactId>log4j-bom</artifactId>
                    <version>2.17.2</version>
                    <type>pom</type>
                    <scope>import</scope>
                  </dependency>
                </dependencies>
              </dependencyManagement>
              <dependencies>
                <dependency>
                  <groupId>org.apache.logging.log4j</groupId>
                  <artifactId>log4j-core</artifactId>
                  <version>2.17.2</version>
                </dependency>
              </dependencies>
            </project>
            """
          )
        );
    }

    @Test
    void propertiesAsGAVCoordinates() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("${quarkus.platform.group-id}", "${quarkus.platform.artifact-id}",
            "${quarkus.platform.version}", "import", "pom", null,null, null, null, null)),
          pomXml(
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>core</artifactId>
              <version>1</version>
              <properties>
                <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
                <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
                <quarkus.platform.version>3.2.3.Final</quarkus.platform.version>
              </properties>
            </project>
            """,
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>core</artifactId>
              <version>1</version>
              <properties>
                <quarkus.platform.artifact-id>quarkus-bom</quarkus.platform.artifact-id>
                <quarkus.platform.group-id>io.quarkus.platform</quarkus.platform.group-id>
                <quarkus.platform.version>3.2.3.Final</quarkus.platform.version>
              </properties>
              <dependencyManagement>
                <dependencies>
                  <dependency>
                    <groupId>${quarkus.platform.group-id}</groupId>
                    <artifactId>${quarkus.platform.artifact-id}</artifactId>
                    <version>${quarkus.platform.version}</version>
                    <type>pom</type>
                    <scope>import</scope>
                  </dependency>
                </dependencies>
              </dependencyManagement>
            </project>
            """
          )
        );
    }

    @Test
    void validation()  {
        AddManagedDependency recipe = new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "latest.release", "import",
          "pom", null, null, null, "org.apache.logging:*", true);
        Validated<Object> validated = recipe.validate();
        assertThat(validated).allMatch(Validated::isValid);
    }

    @Test
    void validationAllowsDashesInOnlyIfUsing()  {
        AddManagedDependency recipe = new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "latest.release", "import",
          "pom", null, null, null, "something-with:dashes-is-ok*", true);
        Validated<Object> validated = recipe.validate();
        assertThat(validated).allMatch(Validated::isValid);
    }

    @Test
    void badCharactersInOnlyIfUsingAreInvalid() {
        AddManagedDependency recipe = new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "latest.release", "import",
          "pom", null, null, null, "spaced group:*", true);
        Validated<Object> validated = recipe.validate();
        assertThat(validated.isValid()).isFalse();
        assertThat(validated.failures()).anyMatch(v -> "onlyIfUsing".equals(v.getProperty()));
    }

    @Test
    void doNotAddIfTypeNotUsed() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "2.17.2", "import",
            "pom", null,null, null, "org.apache.logging.log4j:*", false)),
          pomXml(
            """
            <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
            </project>
            """
          )
        );
    }

    @Test
    void addedToRootPom() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "2.17.2", "import",
            "pom", null,null, null, null, true)),
          pomXml(
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>project</artifactId>
              <version>1</version>
              <modules>
                <module>core</module>
                <module>service</module>
              </modules>
            </project>
            """,
            """
            <project>
              <groupId>com.mycompany.app</groupId>
              <artifactId>project</artifactId>
              <version>1</version>
              <modules>
                <module>core</module>
                <module>service</module>
              </modules>
              <dependencyManagement>
                <dependencies>
                  <dependency>
                    <groupId>org.apache.logging.log4j</groupId>
                    <artifactId>log4j-bom</artifactId>
                    <version>2.17.2</version>
                    <type>pom</type>
                    <scope>import</scope>
                  </dependency>
                </dependencies>
              </dependencyManagement>
            </project>
            """
          ),
          mavenProject("service",
            pomXml(
              """
                <project>
                  <parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>project</artifactId>
                    <version>1</version>
                  </parent>
                  <groupId>com.mycompany.app</groupId>
                  <artifactId>service</artifactId>
                  <version>1</version>
                </project>
              """
            )
          ),
          mavenProject("core",
            pomXml(
              """
                <project>
                  <groupId>com.mycompany.app</groupId>
                  <artifactId>core</artifactId>
                  <version>1</version>
                </project>
              """,
              """
                <project>
                  <groupId>com.mycompany.app</groupId>
                  <artifactId>core</artifactId>
                  <version>1</version>
                  <dependencyManagement>
                    <dependencies>
                      <dependency>
                        <groupId>org.apache.logging.log4j</groupId>
                        <artifactId>log4j-bom</artifactId>
                        <version>2.17.2</version>
                        <type>pom</type>
                        <scope>import</scope>
                      </dependency>
                    </dependencies>
                  </dependencyManagement>
                </project>
              """
            )
          )
        );
    }

    @Test
    void versionSelectionTakesExistingIntoAccount() {
        rewriteRun(
          spec -> spec.recipes(new AddManagedDependency("com.fasterxml.jackson.core", "jackson-databind", "latest.patch", null,
            null, null, null, null, null, true)),
          //language=xml
          pomXml(
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencies>
                  <dependency>
                    <groupId>org.openrewrite</groupId>
                    <artifactId>rewrite-java</artifactId>
                    <version>7.0.0</version>
                  </dependency>
                </dependencies>
              </project>
              """,
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencyManagement>
                  <dependencies>
                    <dependency>
                      <groupId>com.fasterxml.jackson.core</groupId>
                      <artifactId>jackson-databind</artifactId>
                      <version>2.12.7.2</version>
                    </dependency>
                  </dependencies>
                </dependencyManagement>
                <dependencies>
                  <dependency>
                    <groupId>org.openrewrite</groupId>
                    <artifactId>rewrite-java</artifactId>
                    <version>7.0.0</version>
                  </dependency>
                </dependencies>
              </project>
              """
          )
        );
    }

    @Test
    void doesNotAddManagedDependencyIfTransitiveVersionIsTheSameAsRequested() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("com.fasterxml.jackson.core", "jackson-databind", "2.18.0", null,
            null, null, null, null, null, false)),
          pomXml(
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencies>
                  <dependency>
                    <groupId>com.fasterxml.jackson.dataformat</groupId>
                    <artifactId>jackson-dataformat-xml</artifactId>
                    <version>2.18.0</version>
                  </dependency>
                </dependencies>
              </project>
              """
          )
        );
    }

    @Test
    void addsWhenVersionIsHigherThanTransitiveVersion() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("com.fasterxml.jackson.core", "jackson-databind", "2.19.0", null,
            null, null, null, null, null, false)),
          pomXml(
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencies>
                  <dependency>
                    <groupId>com.fasterxml.jackson.dataformat</groupId>
                    <artifactId>jackson-dataformat-xml</artifactId>
                    <version>2.18.0</version>
                  </dependency>
                </dependencies>
              </project>
              """,
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencyManagement>
                  <dependencies>
                    <dependency>
                      <groupId>com.fasterxml.jackson.core</groupId>
                      <artifactId>jackson-databind</artifactId>
                      <version>2.19.0</version>
                    </dependency>
                  </dependencies>
                </dependencyManagement>
                <dependencies>
                  <dependency>
                    <groupId>com.fasterxml.jackson.dataformat</groupId>
                    <artifactId>jackson-dataformat-xml</artifactId>
                    <version>2.18.0</version>
                  </dependency>
                </dependencies>
              </project>
              """
          )
        );
    }

    @Test
    void doesNotDowngradeVersionIfTransitiveDependencyAppearsOnceInTree() {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("com.fasterxml.jackson.core", "jackson-databind", "2.15.3", null,
            null, null, null, null, null, false)),
          pomXml(
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencies>
                  <dependency>
                    <groupId>com.fasterxml.jackson.dataformat</groupId>
                    <artifactId>jackson-dataformat-xml</artifactId>
                    <version>2.18.0</version>
                  </dependency>
                </dependencies>
              </project>
              """
          )
        );
    }

    @ParameterizedTest
    @ValueSource(strings = {"2.14.0", "2.15.3", "2.18.0"})
    void doesNotDowngradeVersionIfTransitiveDependencyAppearsInTreeWithDifferentVersions(String requestedVersion) {
        rewriteRun(
          spec -> spec.recipe(new AddManagedDependency("com.fasterxml.jackson.core", "jackson-databind", requestedVersion, null,
            null, null, null, null, null, false)),
          pomXml(
            """
              <project>
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
                <dependencies>
                  <dependency>
                    <groupId>com.fasterxml.jackson.dataformat</groupId>
                    <artifactId>jackson-dataformat-xml</artifactId>
                    <version>2.18.0</version>
                  </dependency>
                  <dependency>
                    <groupId>com.fasterxml.jackson.datatype</groupId>
                    <artifactId>jackson-datatype-jsr310</artifactId>
                    <version>2.15.3</version>
                  </dependency>
                </dependencies>
              </project>
              """
          )
        );
    }
}

File: rewrite-maven/src/test/kotlin/org/openrewrite/maven/AddManagedDependencyTest.kt
/*
 * Copyright 2021 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.maven

import org.assertj.core.api.Assertions.assertThat
import org.intellij.lang.annotations.Language
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import org.openrewrite.InMemoryExecutionContext
import org.openrewrite.Tree
import org.openrewrite.Validated
import org.openrewrite.java.marker.JavaProject
import org.openrewrite.xml.XmlParser
import java.nio.file.Path

class AddManagedDependencyTest {
    private val xmlParser = XmlParser()
    private val mavenParser = MavenParser.builder().build()

    private val javaProject = JavaProject(Tree.randomId(), "myproject", null)

    @Test
    fun `Recipe validation`()  {
        val recipe = AddManagedDependency("org.apache.logging.log4j", "log4j-bom", "latest.release", "import",
            "pom", null, null, null, "org.apache.logging", true)
        val validated : Validated = recipe.validate()
        assertThat(validated).anyMatch { v ->
            !v.elementAt(0).isValid
        }
    }

    @Test
    fun `Does not use type manage dependency not added`() {
        assertThat(
            addManagedDependency(false,"org.apache.logging.log4j:*")
                .run(
                    mavenParser.parseWithProvenance(
                        """
                        <project>
                            <groupId>com.mycompany.app</groupId>
                            <artifactId>my-app</artifactId>
                            <version>1</version>
                        </project>
                    """.trimIndent()
                    ) + xmlParser.parseWithProvenance(
                        """
                        <project>
                            <groupId>com.mycompany.app</groupId>
                            <artifactId>my-app</artifactId>
                            <version>1</version>
                        </project>
                    """.trimIndent()
                    )
                ).results
        ).isEmpty()
    }

    @Test
    fun `Only added when using is defined and has dependency`() {
        val results = addManagedDependency(false, "org.apache.logging.log4j:*")
                .run(mavenParser.parseWithProvenance(
                        """
                        <project>
                            <groupId>com.mycompany.app</groupId>
                            <artifactId>my-app</artifactId>
                            <version>1</version>
                            <dependencies>
                                <dependency>
                                    <groupId>org.apache.logging.log4j</groupId>
                                    <artifactId>log4j-core</artifactId>
                                    <version>2.17.2</version>
                                </dependency>
                            </dependencies>
                        </project>
                    """.trimIndent()
                    ) + xmlParser.parseWithProvenance(
                        """
                        <project>
                            <groupId>com.mycompany.app</groupId>
                            <artifactId>my-app</artifactId>
                            <version>1</version>
                            <dependencies>
                                <dependency>
                                    <groupId>org.apache.logging.log4j</groupId>
                                    <artifactId>log4j-core</artifactId>
                                    <version>2.17.2</version>
                                </dependency>
                            </dependencies>
                        </project>
                    """.trimIndent()
                    )
                ).results
        assertThat(results).hasSize(1)
        assertThat(results[0].after!!.printAllTrimmed()).isEqualTo(
            """
                    <project>
                        <groupId>com.mycompany.app</groupId>
                        <artifactId>my-app</artifactId>
                        <version>1</version>
                        <dependencyManagement>
                            <dependencies>
                                <dependency>
                                    <groupId>org.apache.logging.log4j</groupId>
                                    <artifactId>log4j-bom</artifactId>
                                    <version>2.17.2</version>
                                    <type>pom</type>
                                    <scope>import</scope>
                                </dependency>
                            </dependencies>
                        </dependencyManagement>
                        <dependencies>
                            <dependency>
                                <groupId>org.apache.logging.log4j</groupId>
                                <artifactId>log4j-core</artifactId>
                                <version>2.17.2</version>
                            </dependency>
                        </dependencies>
                    </project>
                """.trimIndent()
        )
    }


    @Test
    fun `Added to the root pom`(@TempDir tempDir: Path) {
        val project = tempDir.resolve("pom.xml")
        val service = tempDir.resolve("service/pom.xml")
        val core = tempDir.resolve("core/pom.xml")


        service.toFile().parentFile.mkdirs()
        core.toFile().parentFile.mkdirs()

        project.toFile().writeText(
            //language=xml
            """
                <project>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>project</artifactId>
                    <version>1</version>
                    <modules>
                        <module>core</module>
                        <module>service</module>
                    </modules>
                </project>
            """.trimIndent()
        )

        service.toFile().writeText(
            //language=xml
            """
                <project>
                    <parent>
                        <groupId>com.mycompany.app</groupId>
                        <artifactId>project</artifactId>
                        <version>1</version>
                    </parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>service</artifactId>
                    <version>1</version>
                </project>
            """.trimIndent()
        )

        core.toFile().writeText(
            //language=xml
            """
                <project>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>core</artifactId>
                    <version>1</version>
                </project>
            """.trimIndent()
        )
        val results = addManagedDependency(true, null)
            .run(mavenParser.parse(listOf(project, service, core), tempDir, InMemoryExecutionContext())
                            .map { j -> j.withMarkers(j.markers.addIfAbsent(javaProject)) }
                            .mapIndexed { n, maven ->
                                if (n == 0) {
                                    // give the parent a different java project
                                    maven.withMarkers(maven.markers.compute(javaProject) { j, _ -> j.withId(Tree.randomId()) })
                                } else maven
                            }
            ).results
        assertThat(results).hasSize(2)
        assertThat(results[0].after!!.printAllTrimmed()).isEqualTo(
            """
                <project>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>project</artifactId>
                    <version>1</version>
                    <modules>
                        <module>core</module>
                        <module>service</module>
                    </modules>
                    <dependencyManagement>
                        <dependencies>
                            <dependency>
                                <groupId>org.apache.logging.log4j</groupId>
                                <artifactId>log4j-bom</artifactId>
                                <version>2.17.2</version>
                                <type>pom</type>
                                <scope>import</scope>
                            </dependency>
                        </dependencies>
                    </dependencyManagement>
                </project>
            """.trimIndent()
        )
        assertThat(results[1].after!!.printAllTrimmed()).isEqualTo(
            """
                <project>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>core</artifactId>
                    <version>1</version>
                    <dependencyManagement>
                        <dependencies>
                            <dependency>
                                <groupId>org.apache.logging.log4j</groupId>
                                <artifactId>log4j-bom</artifactId>
                                <version>2.17.2</version>
                                <type>pom</type>
                                <scope>import</scope>
                            </dependency>
                        </dependencies>
                    </dependencyManagement>
                </project>
            """.trimIndent()
        )
    }

    @Test
    fun `Added to the root pom having parent not within the source set`(@TempDir tempDir: Path) {
        val parent = tempDir.resolve("pom.xml")
        val child = tempDir.resolve("server/pom.xml")
        child.toFile().parentFile.mkdirs()

        parent.toFile().writeText(
            //language=xml
            """
                <project>
                    <parent>
                        <groupId>org.springframework.boot</groupId>
                        <artifactId>spring-boot-starter-parent</artifactId>
                        <version>2.3.6.RELEASE</version>
                    </parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>my-parent</artifactId>
                    <version>1</version>
                </project>
            """.trimIndent()
        )

        child.toFile().writeText(
            //language=xml
            """
                <project>
                    <parent>
                        <groupId>com.mycompany.app</groupId>
                        <artifactId>my-parent</artifactId>
                        <version>1</version>
                    </parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>my-app</artifactId>
                    <version>1</version>
                </project>
            """.trimIndent()
        )
        val results = addManagedDependency(true, null)
            .run(mavenParser.parse(listOf(parent, child), tempDir, InMemoryExecutionContext())
                            .map { j -> j.withMarkers(j.markers.addIfAbsent(javaProject)) }
                            .mapIndexed { n, maven ->
                                if (n == 0) {
                                    // give the parent a different java project
                                    maven.withMarkers(maven.markers.compute(javaProject) { j, _ -> j.withId(Tree.randomId()) })
                                } else maven
                            }
            ).results
        assertThat(results).hasSize(1)
        assertThat(results[0].after!!.printAllTrimmed()).isEqualTo(
            """
                <project>
                    <parent>
                        <groupId>org.springframework.boot</groupId>
                        <artifactId>spring-boot-starter-parent</artifactId>
                        <version>2.3.6.RELEASE</version>
                    </parent>
                    <groupId>com.mycompany.app</groupId>
                    <artifactId>my-parent</artifactId>
                    <version>1</version>
                    <dependencyManagement>
                        <dependencies>
                            <dependency>
                                <groupId>org.apache.logging.log4j</groupId>
                                <artifactId>log4j-bom</artifactId>
                                <version>2.17.2</version>
                                <type>pom</type>
                                <scope>import</scope>
                            </dependency>
                        </dependencies>
                    </dependencyManagement>
                </project>
            """.trimIndent()
        )
    }

    private fun MavenParser.parseWithProvenance(@Language("xml") vararg pomSources: String) =
        parse(*pomSources).map { j ->
            j.withMarkers(j.markers.addIfAbsent(javaProject))
        }

    private fun XmlParser.parseWithProvenance(@Language("xml") vararg pomSources: String) =
        parse(*pomSources).map { j ->
            j.withMarkers(j.markers.addIfAbsent(javaProject))
        }

    private fun addManagedDependency(
        addToRoot: Boolean,
        onlyIfUsing: String?
    ): AddManagedDependency {
        return AddManagedDependency(
            "org.apache.logging.log4j", "log4j-bom", "2.17.2", "import",
            "pom", null,null, null, onlyIfUsing, addToRoot
        )
    }
}
